app/Console/Commands/AutoWarning.php

```diff
@@ -61,14 +61,16 @@ public function handle(): void
                 ->where('updated_at', '<', $carbon->copy()->subDays(config('hitrun.grace'))->toDateTimeString())
                 ->get();
 
+            $warnings = Warning::withTrashed()
+                ->whereIn('torrent', $hitrun->pluck('torrent.id'))
+                ->whereIn('user_id', $hitrun->pluck('user.id'))
+                ->get();
+
             foreach ($hitrun as $hr) {
                 if (!$hr->user->group->is_immune && $hr->actual_downloaded > ($hr->torrent->size * (config('hitrun.buffer') / 100))) {
-                    $exsist = Warning::withTrashed()
-                        ->where('torrent', '=', $hr->torrent->id)
-                        ->where('user_id', '=', $hr->user->id)
-                        ->first();
+                    $exsist = $warnings->where('torrent', '=', $hr->torrent->id)->firstWhere('user_id', $hr->user->id);
```

Comment:
Add a relation between histories and warnings so you can just do `->with('warning')`